OpenRA.Game/GameRules/WeaponInfo.cs

```diff
@@ -33,6 +33,15 @@ public class ProjectileArgs
 		public Target GuidedTarget;
 	}
 
+	public class WarheadArgs
```

Comment:
I can see why we have this as its own distinct class instead of referencing `ProjectileArgs` directly (we have cases that don't start with a projectile), but having every use case manually copy over the fields adds a lot of duplication and potential headaches in the future if we want add more fields in the future.

We could solve this by adding an optional constructor
```csharp
public WarheadArgs(ProjectileArgs args)
{
	Weapon = args.Weapon;
	DamageModifiers = args.DamageModifiers;
	Source = args.Source;
	SourceActor = args.SourceActor;
	WeaponTarget = args.GuidedTarget;
}
```

and then cases like `AreaBeam` that do have a projectile but want to override with different values can do something like:
```csharp
var warheadArgs = new WarheadArgs(args)
{
	DamageModifiers = adjustedModifiers.ToArray(),
	Source = target
};
```

Reply:
Where would I have to put them? I'm getting a 'cannot put members directly in namespace' compile error if I just put them into WeaponInfo.

Reply:
```csharp
public class WarheadArgs
{
	public WeaponInfo Weapon;
	public int[] DamageModifiers;
	public WPos Source;
	public Actor SourceActor;
	public Target WeaponTarget;

	public WarheadArgs(ProjectileArgs args)
	{
		Weapon = args.Weapon;
		DamageModifiers = args.DamageModifiers;
		Source = args.Source;
		SourceActor = args.SourceActor;
		WeaponTarget = args.GuidedTarget;
	}

	// Default empty constructor for callers that want to initialize fields themselves
	public WarheadArgs() { }
}
```